File: agent/modules/phone.py
# agent/modules/phone.py
import phonenumbers
from phonenumbers import (
    geocoder,
    carrier,
    timezone as pn_timezone,
    number_type,
    PhoneNumberType,
    NumberParseException,
)


NUMBER_TYPE_NAMES = {
    PhoneNumberType.MOBILE: "мобильный",
    PhoneNumberType.FIXED_LINE: "стационарный",
    PhoneNumberType.FIXED_LINE_OR_MOBILE: "стационарный или мобильный",
    PhoneNumberType.TOLL_FREE: "бесплатный (toll-free)",
    PhoneNumberType.PREMIUM_RATE: "премиум",
    PhoneNumberType.SHARED_COST: "shared cost",
    PhoneNumberType.VOIP: "VoIP",
    PhoneNumberType.PERSONAL_NUMBER: "личный",
    PhoneNumberType.PAGER: "пейджер",
    PhoneNumberType.UAN: "UAN",
    PhoneNumberType.VOICEMAIL: "голосовая почта",
    PhoneNumberType.UNKNOWN: "неизвестно",
}


def module_phone(target: str) -> dict:
    target = target.strip()
    result = {"target": target}

    # Парсим номер. Пытаемся угадать регион, если не указан +.
    try:
        parsed = phonenumbers.parse(target, None)
    except NumberParseException as e:
        # Попробуем как российский по умолчанию — часто это самый частый кейс
        try:
            parsed = phonenumbers.parse(target, "RU")
            result["_assumed_region"] = "RU"
        except NumberParseException as e2:
            return {"error": f"не удалось распарсить номер: {e}"}

    # Валидность
    result["is_valid"] = phonenumbers.is_valid_number(parsed)
    result["is_possible"] = phonenumbers.is_possible_number(parsed)

    # Страна
    region_code = phonenumbers.region_code_for_number(parsed)
    result["country_code"] = parsed.country_code
    result["national_number"] = parsed.national_number
    result["region_code"] = region_code

    # Регион (город)
    try:
        region = geocoder.description_for_number(parsed, "ru")
        result["region"] = region or None
    except Exception:
        result["region"] = None

    # Также на английском — иногда русский пустой
    try:
        region_en = geocoder.description_for_number(parsed, "en")
        result["region_en"] = region_en or None
    except Exception:
        result["region_en"] = None

    # Оператор
    try:
        carrier_name = carrier.name_for_number(parsed, "ru")
        if not carrier_name:
            carrier_name = carrier.name_for_number(parsed, "en")
        result["carrier"] = carrier_name or None
    except Exception:
        result["carrier"] = None

    # Тип линии
    try:
        ntype = number_type(parsed)
        result["number_type"] = NUMBER_TYPE_NAMES.get(ntype, "неизвестно")
    except Exception:
        result["number_type"] = "неизвестно"

    # Часовые пояса
    try:
        tzs = pn_timezone.time_zones_for_number(parsed)
        result["timezones"] = list(tzs) if tzs else []
    except Exception:
        result["timezones"] = []

    # Форматы
    try:
        result["formats"] = {
            "international": phonenumbers.format_number(
                parsed, phonenumbers.PhoneNumberFormat.INTERNATIONAL
            ),
            "national": phonenumbers.format_number(
                parsed, phonenumbers.PhoneNumberFormat.NATIONAL
            ),
            "e164": phonenumbers.format_number(
                parsed, phonenumbers.PhoneNumberFormat.E164
            ),
            "rfc3966": phonenumbers.format_number(
                parsed, phonenumbers.PhoneNumberFormat.RFC3966
            ),
        }
    except Exception:
        result["formats"] = {}

    # Геокодер иногда отдаёт "Россия" без города — попробуем ещё раз для мобильных
    if not result.get("region") and region_code:
        try:
            region2 = geocoder.description_for_valid_number(parsed, "ru")
            if region2:
                result["region"] = region2
        except Exception:
            pass

    return result

File: panel/modules/renders/phone.js
// panel/modules/renders/phone.js
import { field } from "../utils.js";

export function renderPhoneResult(r) {
    const parts = [];

    const rows = [];
    rows.push(field("Номер", r.target));
    if (r._assumed_region) {
        rows.push(field("Регион по умолчанию", r._assumed_region, "warn"));
    }

    rows.push(field("Страна (код)", r.country_code ? `+${r.country_code}` : null));
    rows.push(field("Регион (ISO)", r.region_code));

    if (r.is_valid === true) {
        rows.push(field("Валидность", "✅ валидный", "good"));
    } else if (r.is_valid === false) {
        rows.push(field("Валидность", "❌ невалидный", "bad"));
    }

    if (r.is_possible !== undefined) {
        rows.push(field("Возможность", r.is_possible ? "да" : "нет"));
    }

    if (r.region || r.region_en) {
        const reg = r.region || r.region_en;
        const regEn = (r.region_en && r.region_en !== reg) ? ` (${r.region_en})` : "";
        rows.push(field("Гео", `${reg}${regEn}`));
    }

    if (r.carrier) {
        rows.push(field("Оператор", r.carrier));
    }

    if (r.number_type) {
        const cls = r.number_type === "мобильный" ? "good" : "";
        rows.push(field("Тип линии", r.number_type, cls));
    }

    if (r.timezones && r.timezones.length) {
        rows.push(field("Часовые пояса", r.timezones.join(", ")));
    }

    parts.push(`
        <details open>
            <summary>Информация о номере</summary>
            <table class="kv">${rows.join("")}</table>
        </details>
    `);

    const f = r.formats || {};
    if (Object.keys(f).length) {
        parts.push(`
            <details>
                <summary>Форматы</summary>
                <table class="kv">
                    ${field("E.164", f.e164)}
                    ${field("Международный", f.international)}
                    ${field("Национальный", f.national)}
                    ${field("RFC3966", f.rfc3966)}
                </table>
            </details>
        `);
    }

    return parts.join("");
}